### DISCUSSION TITLE
fast-glob/out/readers/stream.js: Unexpected token (16:10)

### DISCUSSION THREAD

**Discussion by mion-kr** [mikro-orm/mikro-orm#2852] [Q&A]
**Describe the bug**
Added mikroORM Package in nestjs, serverlss development environment.

Serverless offline works well, but the following error message occurs when attempting `serverless deploy --stage=development`

Adding only one "@mikro-orm/core": "^5.0.5" in package.json and serverless deployment results in the following error message.

**Stack trace**
```
✖ Stack study-monorepo-serverless-development failed to deploy (32s)
Environment: darwin, node 16.14.0, framework 3.4.0 (local), plugin 6.1.2, SDK 4.3.1
Credentials: Local, "default" profile
Docs:        docs.serverless.com
Support:     forum.serverless.com
Bugs:        github.com/serverless/serverless/issues

Error:
Error: Parsing file /Users/jongyoon-kim/Dropbox/02_Develop/01_KJY/10. Nest/02. toy-project/toy-estate-api/.build/node_modules/fast-glob/out/readers/stream.js: Unexpected token (16:10)
    at Deps.parseDeps (/Users/jongyoon-kim/Dropbox/02_Develop/01_KJY/10. Nest/02. toy-project/toy-estate-api/node_modules/module-deps/index.js:519:15)
    at getDeps (/Users/jongyoon-kim/Dropbox/02_Develop/01_KJY/10. Nest/02. toy-project/toy-estate-api/node_modules/module-deps/index.js:447:44)
    at /Users/jongyoon-kim/Dropbox/02_Develop/01_KJY/10. Nest/02. toy-project/toy-estate-api/node_modules/module-deps/index.js:430:38
    at ConcatStream.<anonymous> (/Users/jongyoon-kim/Dropbox/02_Develop/01_KJY/10. Nest/02. toy-project/toy-estate-api/node_modules/concat-stream/index.js:37:43)
    at ConcatStream.emit (node:events:532:35)
    at ConcatStream.emit (node:domain:475:12)
    at finishMaybe (/Users/jongyoon-kim/Dropbox/02_Develop/01_KJY/10. Nest/02. toy-project/toy-estate-api/node_modules/concat-stream/node_modules/readable-stream/lib/_stream_writable.js:630:14)
    at endWritable (/Users/jongyoon-kim/Dropbox/02_Develop/01_KJY/10. Nest/02. toy-project/toy-estate-api/node_modules/concat-stream/node_modules/readable-stream/lib/_stream_writable.js:638:3)
    at ConcatStream.Writable.end (/Users/jongyoon-kim/Dropbox/02_Develop/01_KJY/10. Nest/02. toy-project/toy-estate-api/node_modules/concat-stream/node_modules/readable-stream/lib/_stream_writable.js:594:41)
    at DuplexWrapper.onend (/Users/jongyoon-kim/Dropbox/02_Develop/01_KJY/10. Nest/02. toy-project/toy-estate-api/node_modules/duplexer2/node_modules/readable-stream/lib/_stream_readable.js:577:10)
    at Object.onceWrapper (node:events:639:28)
    at DuplexWrapper.emit (node:events:532:35)
    at DuplexWrapper.emit (node:domain:475:12)
    at endReadableNT (/Users/jongyoon-kim/Dropbox/02_Develop/01_KJY/10. Nest/02. toy-project/toy-estate-api/node_modules/duplexer2/node_modules/readable-stream/lib/_stream_readable.js:1010:12)
    at processTicksAndRejections (node:internal/process/task_queues:83:21)

```

**To Reproduce**
[my gitlab](https://gitlab.com/kjy-study/nest/pre/toy-project/estate/toy-estate-api)
> You have to prepare the key and secret of AWS IAM
1. npm i -g serverless
2. npm i
3. sls config credentials --provider aws --key ${aws_access_key_id} --secret ${aws_secret_access_key}
4. sls deploy --stage=development

**Expected behavior**
```
✔ Service deployed to stack study-monorepo-serverless-development (195s)

endpoints:
  ANY - https://abcd.ap-northeast-1.amazonaws.com/development/service-a/{any+}
  ANY - https://abcd.ap-northeast-1.amazonaws.com/development/service-a
  ANY - https://abcd.ap-northeast-1.amazonaws.com/development/service-b/{any+}
  ANY - https://abcd.ap-northeast-1.amazonaws.com/development/service-b
  ANY - https://abcd.ap-northeast-1.amazonaws.com/development/service-c/{any+}
  ANY - https://abcd.ap-northeast-1.amazonaws.com/development/service-c
functions:
  service-a: study-monorepo-serverless-development-service-a (22 MB)
  service-b: study-monorepo-serverless-development-service-b (22 MB)
  service-c: study-monorepo-serverless-development-service-c (22 MB)


```
**Additional context**

**Versions**

| Dependency | Version |
| - | - |
| node | v16.14.0 |
| typescript | ^4.3.5 |
| mikro-orm | ^5.0.5 |
| your-driver | ? |
| os | 2020 M1 macbook |

**Comment by B4nan** [comment]
Nothing in the stack trace is pointing to the ORM 🤷 

Cant help with this, if you want to report bugs, provide reproduction only with the ORM involved.

**Reply by mion-kr** [reply]
@B4nan 

The above error is related to ORM. 
Error stack doesn't point to ORM. but package.json is using the 'globby' package.

I re-created the serverless project and added the MikroORM package. And I did serverless deployment again. However, the fast-glob/out/readers/stream.js error occurs the same.

> [serverless-mikroorm project](https://gitlab.com/kjy-study/nest/pre/example/error-serverless-mikroorm) 

I think this issue should be treated as an issue. This is because in the nestjs serverless environment, it is not deployed due to mikroORM.

**Reply by B4nan** [reply]
If you think it's an issue with globby integration, provide a repro that proves it - by doing only `MikroORM.init()` without any other dependency. Your "repro" is quite huge project that combines many technologies and even AWS cloud. Reproductions need to be minimal, this sounds more like you want me to help you with your project, hence I moved it to GH discussions. I will be happy to help, but don't expect me to keep this as an issue unless we have a clear reproduction that points to a bug, and don't expect me to learn technologies I never used just to be able to help you. I am here to help with the ORM usage and its configuration, can't help with AWS and serverless, as I never really used it.

My guess is that you are trying to compile things into single file - but you can't use folder based discovery in such case, as there are no folders, just one file.

**Reply by mion-kr** [reply]
@B4nan 

I didn't mean to ask you to help me with my serverless project. I just wanted to let you know that adding the MikroORM package in a Serverless environment will cause problems. When I started NestJS, I was fascinated by MikroORM. I added MikroORM to the serverless project with a swollen dream.

I'll look for another way.
Thank you for your answer.

**Reply by B4nan** [reply]
I briefly checked your repro again, you dont use folder based discovery, which make me even more sure this is not coming from the ORM. There are very few places where we use globby, basically just folder based discovery and metadata cache handling with `FileCacheAdapter`, and your repro is apparently not using any of those.

The repro looks like you are trying to bundle fast-glob and the compiled version is invalid, have you checked what is in that problematic file on line 16? 

Maybe try to install fast-glob and use that (and remove everything ORM related), I can imagine that itself could cause the failure.

**Reply by martintanbit** [reply]
I had this exact same issue and for my case I was able to solve it (or workaround it) by disabling `serverless-plugin-optimize` in my `serverless.yaml`. It seems that this issue has nothing really to do with MikroORM or `fast-glob` and more to do with a missing configuration for `serverless-plugin-optimize` for the `static` keyword in `fast-glob/out/readers/stream.js` line 16.

I don't have much experience with configuration for transpiling but I'm guessing we have to add some Babel plugin to fix that `static` reserved keyword issue.

**Comment by huangdaojian** [comment]
谢谢~已收到！！！！！！                              祝你事事顺利，天天开心
